auth/auth/auth.py

```diff
@@ -50,6 +63,60 @@ async def user_from_login_id(db, login_id):
     return None
 
 
+async def user_from_username(db, username):
+    users = [x async for x in db.select_and_fetchall("SELECT * FROM users WHERE username = %s;", username)]
+    if len(users) == 1:
+        return users[0]
+    assert len(users) == 0, users
+    return None
+
+
+async def users_with_username_or_login_id(db: Database, username: str, login_id: Optional[str]) -> List[dict]:
+    where_conditions = ['username = %s']
+    where_args = [username]
+
+    if login_id is not None:
+        where_conditions.append('login_id = %s')
+        where_args.append(login_id)
+
+    existing_users = [
+        x
+        async for x in db.select_and_fetchall(f"SELECT * FROM users WHERE {' OR '.join(where_conditions)};", where_args)
+    ]
+
+    return existing_users
+
+
+async def check_valid_new_user(db: Database, username, login_id, is_developer, is_service_account):
+    if not isinstance(username, str):
+        raise InvalidType('username', username, 'str')
+    if login_id is not None and not isinstance(login_id, str):
+        raise InvalidType('login_id', login_id, 'str')
+    if not isinstance(is_developer, bool):
+        raise InvalidType('is_developer', is_developer, 'bool')
+    if not isinstance(is_service_account, bool):
+        raise InvalidType('is_service_account', is_service_account, 'bool')
+    if is_developer and is_service_account:
+        raise MultipleUserTypes(username)
+    if not is_service_account and not login_id:
+        raise EmptyLoginID(username)
+    if not username or not all(c for c in username if c.isalnum()):
+        raise InvalidUsername(username)
+
+    existing_users = await users_with_username_or_login_id(db, username, login_id)
+
+    if len(existing_users) == 1:
+        existing_user = existing_users[0]
+        expected_username = existing_user['username']
+        expected_login_id = existing_user['login_id']
+        if username != expected_username:
+            raise DuplicateLoginID(username, login_id)
```

Comment:
In the message you claim that the username is the username of the already extant user, so I think this should be:

```
raise DuplicateLoginID(existing_user['username'], login_id)
```